src/session_impl.cpp

```diff
@@ -5135,14 +5113,15 @@ namespace aux {
 #if TORRENT_USE_ASSERTS
 		sha1_hash i_hash = t.torrent_file().info_hash();
 #endif
-#ifndef TORRENT_DISABLE_DHT
-		if (i == m_next_dht_torrent)
-			++m_next_dht_torrent;
-#endif
-		if (i == m_next_lsd_torrent)
-			++m_next_lsd_torrent;
 
-		m_torrents.erase(i);
+		auto array_iter = std::find(m_torrents.begin(), m_torrents.end(), i->second);
+		TORRENT_ASSERT(array_iter != m_torrents.end());
+		if (array_iter != m_torrents.end())
```

Comment:
actually, this should be `array_iter != m_torrents.end() - 1`, right?

Reply:
I don't think it's OK to self-swap

Reply:
You're right, self-swap isn't safe here. We still wan to check for `array_iter != m_torrents.end()` though, because swapping with `end()` is definitely not safe and we don't want to pop some random torrent out of the vector. The `array_iter != m_torrents.end() - 1` check should be nested inside the current check and only cover the call to `std::swap`.

Reply:
Oh, i was under the impression the iterator cannot be end(), given the assert

Reply:
Right, but asserts aren't typically enabled in release builds so I added the check to prevent a crash if the condition arises in the field. Perhaps that's a bit much.